[bugfix] Fix double precision I/O in form widgets

elpaso opened the pull request:
## Fix some display inconsistencies with double and integer numbers in forms:

- range widget and text input now display numbers in the same way 
- range widget and text input now display numbers respecting system settings
- accept comma or dot according to locale settings  (on unix it should accept both, on windows only the decimal point reported by the system settings)

Try to fix  https://issues.qgis.org/issues/18963, https://issues.qgis.org/issues/18679

There is a windows build here if someone wants to give this patch a quick try:
http://www.itopen.it/bulk/mxe-release-2018-06-05-18-06-30.zip

elpaso committed:
Use QLocale when representing double fields

elpaso committed:
Use displayString to get the formatted value

elpaso committed:
Try to convert doubles with comma as decimal point

elpaso committed:
Accept dot and comma as decimal point on input

elpaso commented:
@m-kuhn , @nyalldawson this PR introduces some core changes in number formatting, can you please have a look?
There are some other related bugs ( https://issues.qgis.org/issues/18656 ) I'd like to address if the approach in this PR is accepted.

nyalldawson commented on src/core/qgsfield.cpp:
```diff
@@ -208,8 +209,30 @@ QString QgsField::displayString( const QVariant &v ) const
+    }
+    else
+    {
+      // Precision is not set, let's guess it from the
```
I'm concerned about the performance implications of this block - could we bypass it for "normal" locales and just return v.toString() instead?

elpaso replied:
Do you mean "." == "normal"?

nyalldawson replied:
Yeah, of course ;) Comma is just weird.

elpaso replied:
@nyalldawson I don't think we can: if we go for `toString` we'll miss the thousands separator (we get `123456789.0123` instead of `123'456'789.0123`  (or whatever the system's locale settings say it should be). If I'm not wrong, the range widget formatter does that already and (among other issues) this PR tries to make the numbers look the same whether the range widget of the plain text widget is used.

Do you have any reason to believe that it would make a noticeable performance difference?

m-kuhn replied:
could we bypass this for "normal" locales where "normal" is defined as `.` for decimal separator and nothing for the thousands separator?

elpaso replied:
Well, yes, but if we do, we'll loose the thousand grouping for `.` locales and we get `123456789.0123` instead of `123'456'789.0123` (or whatever the system's locale settings say it should be).
```
In [8]: QLocale(QLocale.German).toString( 123456789.012345, 'f', 6 )
Out[8]: '123.456.789,012345'

In [9]: QLocale(QLocale.English).toString( 123456789.012345, 'f', 6 )
Out[9]: '123,456,789.012345'
```

m-kuhn replied:
I think you are missing the second condition of my definition of normal.

elpaso replied:
Ah, yes, sorry. Ok then.

nyalldawson commented on src/core/qgsfield.cpp:
```diff
@@ -258,6 +281,13 @@ bool QgsField::convertCompatible( QVariant &v ) const
     return true;
   }
 
+  // Give it a chance to convert to double since we accept both comma and dot as decimal point
```
I think this should only be applied for locales which use comma as decimal... it may have odd implications for dot as decimal locales with thousand separated strings

nyalldawson commented on src/gui/qgsfieldvalidator.cpp:
```diff
@@ -56,7 +56,7 @@ QgsFieldValidator::QgsFieldValidator( QObject *parent, const QgsField &field, co
       if ( mField.length() > 0 && mField.precision() > 0 )
       {
-        QString re = QStringLiteral( "-?\\d{0,%1}(\\.\\d{0,%2})?" ).arg( mField.length() - mField.precision() ).arg( mField.precision() );
+        QString re = QStringLiteral( "-?\\d{0,%1}([\\.,]\\d{0,%2})?" ).arg( mField.length() - mField.precision() ).arg( mField.precision() );
```
I think this should be based on locale setting again - or someone entering "1,000" will get a value of "1.0", when we should be rejecting the string.

elpaso replied:
yes, I though about that, the use case is a user entering a floating point, I doubt that anyone would enter a floating point number using the thousand grouping but I might be wrong.
On the other hand there is a clear advantage in accepting both comma and dots for locales that use a comma because many applications accept both and internally convert to the correct format.
I think we could probably put it in a branch for the cases when the decimalPoint is a comma as you suggested to do in the previous comment.

m-kuhn replied:
What about the case where floats with thousand separator are entered through copy-paste?

Wouldn't the following code respect both, always allow `.` and only allow `,` when it's an option?
```
QString re = QStringLiteral( "-?\\d{0,%1}([\\.%2]\\d{0,%3})?" ).arg( mField.length() - mField.precision(), decimalSeparator, mField.precision() );
```

elpaso replied:
:+1: I was going to address this, your proposal sound perfect to me.

elpaso committed:
Set locale in the tests

elpaso commented:
@m-kuhn would you please have a look to this PR? I'll address some of Nyall's comments after yours.

elpaso committed:
Added tests for locales other than english

elpaso committed:
Respect locale decimal separator in input fields

elpaso committed:
Skip custom handling of decimal point if it's a dot

elpaso committed:
Fix test comparison

elpaso committed:
Check if conversion was successful before passing to locale

elpaso committed:
Fix test, because we only accept comma for compatible locales

DelazJ commented:
Also, when updating the values in the form view using the dot, values are displayed with dot until you select another feature and get back to the updated one's form. It's not a real issue given that the values are finally shown with comma but given that in the table view losing the focus on the field automatically updates it with comma, I wanted to report it.

elpaso commented:
@DelazJ thank you so much for testing, this is really tricky because the code follows different paths depending on field width and precision, moreover `QDoubleValidator` on windows do not accept `.` when decimal point is `,`, while on Linux it does. I suspect that this is the cause of your observation about not being able to enter  `.` on certain fields only (can you please check in field properties that precision and width are both `0` for those fields?). If I am right, you should be able to enter a `.` if the fields have either width or precision different than `0`.

I'm hesitant to replace the default `QDoubleValidator` with something more intelligent for windows only and I'd rather prefer to live with the fact that (on windows only and only when both width and precision are `0`) the users must enter theirs locale's defined decimal point (`,` in your case), but I'll see if I can fix it.

I also noticed the `NULL` issue and I'll have a look at it, I think that it is the only blocker before merging this patch or do I miss something important?

Btw, I'll upload another build for testing, thanks again!

elpaso committed:
Use QLocale to parse string representation of numbers

Because we now might have thousand separators, also
fixes the NULL issue reported by DelazJ.

Comes with many test cases.

elpaso commented:
@DelazJ the last commit should fix the NULL issue, here is the new build: http://www.itopen.it/bulk/mxe-release-2018-06-10-16-04-55.zip

nyalldawson commented:
@elpaso while you're on this issue, is there any value in cherry-picking https://github.com/qgis/QGIS/pull/4837/commits/0ba36acf93353ff177f4e149215dbee8b9d5d78d ?

elpaso commented:
@luipir I wouldn't call it a bug, Linux implementation is apparently just smarter.

elpaso committed:
[bugfix] Fix Z&M edition. Fix double conversion from locale.

Mnually Cherry-picked from 0ba36acf93353ff177f4e149215dbee8b9d5d78d

elpaso commented:
AFAIK the only pending issue is the inability for locales that:

do not use a dot as a decimal place AND are on windows AND  field width is 0 AND field precision is 0 AND are entering  double .... 

to use a dot to enter double precision (real, floats etc) numbers

m-kuhn commented:
Would it be a possibility to just bump up the field width to a reasonably high number (more than the 17 significant double digits) in this scenario to workaround this problem?

elpaso commented:
@m-kuhn I think this is not a big issue because the user will still be able to enter the doubles using the decimal separator of his locale.
I think that the definitive solution here would be to create a `QgsDoubleValidator` class that tries first to parse the double using user's locale and if (and only if, on windows) it fails and that locale's decimal point is not a 'dot' then try to parse the double using the english locale.
It shouldn't be hard to to but I'd like to address it at a later point, hopefully after some feedback about this current PR.

m-kuhn commented:
I was just trying to come up with an easier solution than that, but since I don't even completely understand the problem, I can't comment too much on it, never mind ;)
